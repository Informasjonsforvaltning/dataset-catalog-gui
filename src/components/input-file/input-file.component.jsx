import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import './input-file.scss';

export const InputFile = props => {
  const { className, onChange, children } = props;

  return (
    <label className={`inputfile ${className}`} htmlFor="file">
      <input className="inputfile" id="file" type="file" onChange={onChange} />
      {children}
    </label>
  );
};

InputFile.defaultProps = {
  className: null,
  onChange: _.noop
};

InputFile.propTypes = {
  className: PropTypes.string,
  onChange: PropTypes.func,
  children: PropTypes.node.isRequired
};
