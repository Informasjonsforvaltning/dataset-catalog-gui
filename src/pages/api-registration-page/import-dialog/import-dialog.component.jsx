import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'reactstrap';
import { withState, withHandlers, compose } from 'recompose';
import _ from 'lodash';
import { readAsText } from 'promise-file-reader';

import localization from '../../../services/localization';
import { patchApi } from '../../../services/api/registration-api/apis';
import { InputFile } from '../../../components/input-file/input-file.component';
import { ImportLinkUpload } from '../import-link-upload/import-link-upload.component';

export const ImportDialogPure = ({
  catalogId,
  apiId,
  showLinkImport,
  onToggleShowLinkImport,
  handleShowImportSpecificationButtons,
  onSetImportUrl,
  handleShowImportError,
  handleShowImportSuccess,
  importUrl,
  apiSuccess
}) => {
  const onCancel = () => {
    handleShowImportSpecificationButtons();
    handleShowImportError(false);
  };

  const onLinkUpload = () => {
    onToggleShowLinkImport();
    handleShowImportSuccess(false);
    handleShowImportError(false);

    patchApi(catalogId, apiId, { apiSpecUrl: importUrl })
      .then(responseData => {
        apiSuccess(responseData);
        handleShowImportSuccess(true);
        handleShowImportSpecificationButtons();
      })
      .catch(error => {
        if (process.env.NODE_ENV !== 'production') {
          console.log('error', error); // eslint-disable-line no-console
        }
        handleShowImportError(true);
      });
    window.scrollTo(0, 0);
  };

  const onFileUpload = e => {
    handleShowImportSuccess(false);
    handleShowImportError(false);

    const fileMetaData = e.target.files[0] || null;

    if (fileMetaData) {
      readAsText(fileMetaData)
        .then(apiSpec => {
          patchApi(catalogId, apiId, { apiSpec })
            .then(responseData => {
              apiSuccess(responseData);
              handleShowImportSuccess(true);
              handleShowImportSpecificationButtons();
            })
            .catch(error => {
              if (process.env.NODE_ENV !== 'production') {
                console.log('error', error); // eslint-disable-line no-console
              }
              handleShowImportError(true);
            });
        })
        .catch(error => {
          if (process.env.NODE_ENV !== 'production') {
            console.log('error', error); // eslint-disable-line no-console
          }
          handleShowImportError(true);
        });
    }
    window.scrollTo(0, 0);
  };

  if (!showLinkImport) {
    return (
      <>
        <Button
          className="mr-3 fdk-button"
          color="primary"
          onClick={onToggleShowLinkImport}
        >
          {localization.api.import.importLink}
        </Button>
        <InputFile
          className="fdk-button btn btn-primary m-0"
          onChange={onFileUpload}
        >
          {localization.api.import.importFile}
        </InputFile>
        <Button
          testid="button-cancel"
          className="fdk-color-link fdk-bg-color-white"
          color="transparent"
          onClick={onCancel}
        >
          {localization.api.import.cancel}
        </Button>
      </>
    );
  }
  return (
    <ImportLinkUpload
      handleLinkUpload={onLinkUpload}
      handleCancelImport={handleShowImportSpecificationButtons}
      handleChangeImportUrl={onSetImportUrl}
    />
  );
};

ImportDialogPure.defaultProps = {
  catalogId: null,
  apiId: null,
  handleShowImportSpecificationButtons: _.noop,
  importUrl: null,
  handleShowImportError: _.noop,
  handleShowImportSuccess: _.noop,
  showLinkImport: false,
  onToggleShowLinkImport: _.noop,
  onSetImportUrl: _.noop
};

ImportDialogPure.propTypes = {
  catalogId: PropTypes.string,
  apiId: PropTypes.string,
  handleShowImportSpecificationButtons: PropTypes.func,
  importUrl: PropTypes.string,
  handleShowImportError: PropTypes.func,
  handleShowImportSuccess: PropTypes.func,
  showLinkImport: PropTypes.bool,
  onToggleShowLinkImport: PropTypes.func,
  onSetImportUrl: PropTypes.func,
  apiSuccess: PropTypes.func.isRequired
};

const enhance = compose(
  withState('showLinkImport', 'toggleShowLinkImport', false),
  withState('importUrl', 'setImportUrl', ''),
  withHandlers({
    onToggleShowLinkImport: props => e => {
      if (e) {
        e.preventDefault();
      }
      props.toggleShowLinkImport(!props.showLinkImport);
    },
    onSetImportUrl: props => e => {
      e.preventDefault();
      props.setImportUrl(e.target.value);
    }
  })
);

export const ImportDialogPureWithState = enhance(ImportDialogPure);
export const ImportDialog = ImportDialogPureWithState;
