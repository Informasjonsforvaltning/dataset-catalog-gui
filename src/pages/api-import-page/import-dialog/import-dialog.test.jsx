import React from 'react';
import { mount, shallow } from 'enzyme';
import { ImportDialog } from './import-dialog.component';
import { InputFile } from '../../../components/input-file/input-file.component';

let defaultProps;
let wrapper;
let preventDefault;
let showLinkImport;
let handleFileUpload;

beforeEach(() => {
  showLinkImport = jest.fn();
  handleFileUpload = jest.fn();
  preventDefault = jest.fn();
  defaultProps = {
    showLinkImport,
    handleFileUpload
  };
  wrapper = shallow(<ImportDialog {...defaultProps} />);
});

test('should render ImportDialog correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle button show link upload', () => {
  const mockedEvent = {
    preventDefault
  };
  wrapper.find('Button').prop('onClick')(mockedEvent);
  expect(showLinkImport).toHaveBeenCalled();
});

test('should handle input file upload', () => {
  const mockedEvent = {
    preventDefault
  };
  mount(<ImportDialog {...defaultProps} />)
    .find(InputFile)
    .prop('onChange')(mockedEvent);
  expect(handleFileUpload).toHaveBeenCalled();
});
