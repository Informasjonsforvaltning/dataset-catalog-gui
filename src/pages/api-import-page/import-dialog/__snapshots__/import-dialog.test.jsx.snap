// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ImportDialog correctly 1`] = `
<Component>
  <div
    className="mb-5"
  >
    <h3>
      Importer API-spesifikasjon
    </h3>
  </div>
  <div
    className="mb-5"
  >
    Felles datakatalog støtter foreløpig Open API-Specification 2.0 og 3.0 og Swagger 2.0 i JSON format.
  </div>
  <Button
    className="mr-3"
    color="primary"
    onClick={[MockFunction]}
    tag="button"
  >
    Importer via lenke
  </Button>
  <Component
    className="btn btn-primary m-0"
    onChange={[MockFunction]}
  >
    Importer via fil
  </Component>
</Component>
`;
